Document 3.13, 3.14, 4.0 and future removals (#93986)
python/cpython: Doc/whatsnew/3.12.rst

* :mod:`uu`
* :mod:`xdrlib`

Other modules:

* :mod:`lib2to3`, and the ``2to3`` program (:gh:`84540`)

APIs:

* :class:`configparser.LegacyInterpolation` (:gh:`90765`)
* :func:`locale.getdefaultlocale` (:gh:`90817`)
* :meth:`turtle.RawTurtle.settiltangle` (:gh:`50096`)
* :func:`unittest.findTestCases` (:gh:`50096`)
* :func:`unittest.getTestCaseNames` (:gh:`50096`)
* :func:`unittest.makeSuite` (:gh:`50096`)
* :meth:`unittest.TestProgram.usageExit` (:gh:`67048`)
* :class:`webbrowser.MacOSX` (:gh:`86421`)

Pending Removal in Python 3.14
==============================

The following modules and APIs have been deprecated in earlier Python releases,
and will be removed in Python 3.14.

APIs:

*  :meth:`asyncio.Future.cancel`'s *msg* parameter (:gh:`90985`)
*  :meth:`asyncio.Task.cancel`'s *msg* parameter (:gh:`90985`)

> Pending Removal in Python 4.0

Reviewer: Can we remove this in a `3.xy` release? gh-57281 is from 2012 and beliefs about Python 4 were different a decade ago.

Author: That sounds sensible. Please could you create a new issue? It doesn't currently raise a deprecation warning, so if we add one in 3.12 we could remove it in 3.14.

I only find a few uses in the top 5,000 PyPI projects (with regex `"array\(['\"]u['\"]"`):
```python
serpent-1.40/tests/test_serpent.py: ser = serpent.dumps(array.array('u', str("unicode")))
serpent-1.40/tests/test_serpent.py: ser = serpent.dumps(array.array('u', "normal"))
Pyro4-4.82/tests/PyroTests/test_serialize.py: a1 = array.array('u', unicode("hello"))
Pyro4-4.82/tests/PyroTests/test_serialize.py: a1 = array.array('u', unicode("hello"))
numcodecs-0.9.1/numcodecs/tests/test_compat.py: a = array.array('u', 'qwertyuiqwertyui')
pyodbc-4.0.32/setup.py: UNICODE_WIDTH = array('u').itemsize
pyobjc-framework-Cocoa-8.5/PyObjCTest/test_cfstring.py: b = array.array("u", "hello world")
pyobjc-framework-Cocoa-8.5/PyObjCTest/test_cfstring.py: b2 = array.array("u", " " * 40)
nibabel-3.2.2/doc/source/notebooks/cross_product_error.ipynb: "u = Matrix(symarray('u', (3, 1)))\n",
pyenchant-3.2.2/enchant/checker/__init__.py: self._text = array.array("u")
pyenchant-3.2.2/enchant/checker/__init__.py: self._text = array.array("u", text)
biopython-1.79/Bio/Seq.py: return array.array("u", self._data.decode("ASCII"))
biopython-1.79/Tests/test_seq.py: mutable_s = Seq.MutableSeq(array.array("u", sequence.decode("ASCII")))
```

(For reference, here's the commit where it was deprecated in 3.3: https://github.com/python/cpython/commit/029780be09ab5483ac4562d795ea4b4b6d456f44)

Author: Ah, there's already an issue here: https://github.com/python/cpython/issues/80480

* :mod:`uu`
* :mod:`xdrlib`

Other modules:

* :mod:`lib2to3`, and the ``2to3`` program (:gh:`84540`)

APIs:

* :class:`configparser.LegacyInterpolation` (:gh:`90765`)
* :func:`locale.getdefaultlocale` (:gh:`90817`)
* :meth:`turtle.RawTurtle.settiltangle` (:gh:`50096`)
* :func:`unittest.findTestCases` (:gh:`50096`)
* :func:`unittest.getTestCaseNames` (:gh:`50096`)
* :func:`unittest.makeSuite` (:gh:`50096`)
* :meth:`unittest.TestProgram.usageExit` (:gh:`67048`)
* :class:`webbrowser.MacOSX` (:gh:`86421`)

Pending Removal in Python 3.14
------------------------------

The following have been deprecated in earlier Python releases,
and will be removed in Python 3.14.

Classes:

> * Deprecated the following :mod:`importlib.abc` classes:

Reviewer 2: ```suggestion
* The following :mod:`importlib.abc` classes:
```

  (Contributed by Jason R. Coombs and Hugo van Kemenade in :gh:`93963`.)

APIs:

*  :meth:`asyncio.Future.cancel`'s *msg* parameter (:gh:`90985`)
*  :meth:`asyncio.Task.cancel`'s *msg* parameter (:gh:`90985`)

Pending Removal in Python 4.0
-----------------------------

The following have been deprecated in earlier Python releases,

> and will be removed in Python 4.0.

Reviewer 2: Do we need to clarify what we mean with Python 4?

This seems to imply that
1. there will be a Python 4.0;
2. this API will be removed in Python 4.0

However my understanding is that "Python 4" means "we are not going to remove it any time soon, and if we are going to remove it all it will be in the distant future if/when Python 4 is released".

Reviewer: Yup, we shouldn't document something about Python 4 when that's hypothetical and devoid of specific meaning.

If we want to raise DeprecationWarning before we document a 3.xy deprecation timeline (as suggested in https://github.com/python/cpython/pull/93986#discussion_r902734305 ), I'd vote we not document this deprecation until then.

Reviewer 2: The deprecation itself should be documented, and I think it would be better to also document that the removal is not currently planned (exact wording TBD).

Author: In https://github.com/python/cpython/issues/80480, the removal was suggested for "4.0 or 3.10", when we didn't know which was coming next (https://github.com/python/cpython/issues/80480#issuecomment-1093818166):

> The doc says:
> 
> 'u' will be removed together with the rest of the Py_UNICODE API.
Deprecated since version 3.3, will be removed in version 4.0.
[docs.python.org/3/library/array.html](https://docs.python.org/3/library/array.html)
> 
> But DeprecationWarning is not raised yet. Let's raise it.
> 
> * 3.8 -- PendingDeprecationWarning
> * 3.9 -- DeprecationWarning
> * 4.0 or 3.10 -- Remove it.

And https://github.com/python/cpython/issues/80480#issuecomment-1093818167:

> '4.0' is a stand-in for 'sometime after 2.7.final', scheduled for Jan 2020. A Pending... for 3.8.0, scheduled for Oct 2019, seems reasonable to me. Perhaps we should have a pydev discussion for the general issue of post 2.7 removals of already deprecated items.

Please could one of you bump that issue with a suggestion on how to proceed? Maybe it can be removed in 3.12?

---

This PR is documenting the status quo, repeating the 4.0 removal already documented at https://docs.python.org/3/library/array.html to bring more visibility. (This discussion shows it's working! :)

> The deprecation itself should be documented, and I think it would be better to also document that the removal is not currently planned (exact wording TBD).

Agreed. For now, shall we move this to either of these sections?

* the "Pending Removal in Future Versions" section: "The following were deprecated in earlier Python versions and will be removed, although there is currently no date scheduled for their removal."

* or the top-level "Deprecated" section, with no intro text. There's `typing` deprecations here, which may never be removed.

And do we want to delete the 4.0 removal at https://docs.python.org/3/library/array.html?

Reviewer 2: > Please could one of you bump that issue with a suggestion on how to proceed? Maybe it can be removed in 3.12?

I left a comment in #80480.  It seems the code doesn't raise a deprecation warning, so it will have to go through a deprecation cycle before it can be removed.  Once the warning is added, we will know in which release it will be removed and we will be able to update the doc accordingly.

> the "Pending Removal in Future Versions" section:

In the meanwhile I think it can be moved here.

Author: Moved!

  (Contributed by Jason R. Coombs and Hugo van Kemenade in :gh:`93963`.)

APIs:

> *  :meth:`asyncio.Future.cancel`'s *msg* parameter (:gh:`90985`)
> *  :meth:`asyncio.Task.cancel`'s *msg* parameter (:gh:`90985`)

Reviewer 2: ```suggestion
* :meth:`asyncio.Future.cancel`'s *msg* parameter (:gh:`90985`)
* :meth:`asyncio.Task.cancel`'s *msg* parameter (:gh:`90985`)
```

  :keyword:`for`, :keyword:`if`, :keyword:`in`, :keyword:`is` and :keyword:`or`.
  In a future release it will be changed to a syntax error. (:gh:`87999`)

> * :func:`asyncio.get_event_loop` emits a deprecation warning if there is

Reviewer 2: ```suggestion
* :func:`asyncio.get_event_loop` emits a :exc:`DeprecationWarning` if there is
```

  :keyword:`for`, :keyword:`if`, :keyword:`in`, :keyword:`is` and :keyword:`or`.
  In a future release it will be changed to a syntax error. (:gh:`87999`)

> * :func:`asyncio.get_event_loop` emits a :exc:`DeprecationWarning` if there is

Reviewer 3: The deprecation plans for this function have changed. See https://github.com/python/cpython/pull/100970 for the updated deprecation notice.

Author: Thanks for the note! This PR has got pretty out of date...
